weave/scorers/scorer_types.py

```diff
@@ -0,0 +1,274 @@
+from typing import TYPE_CHECKING, Any, Literal, Optional, Union
+
+from pydantic import Field, PrivateAttr, field_validator
+
+import weave
+from weave.scorers.utils import set_device
+
+if TYPE_CHECKING:
+    import torch
+    from transformers.modeling_utils import PreTrainedModel
+    from transformers.pipelines.base import Pipeline
+    from transformers.tokenization_utils import PreTrainedTokenizer
+
+
+class LLMScorer(weave.Scorer):
+    """Score model outputs using a Large Language Model (LLM).
+
+    This scorer leverages LLMs to evaluate and score model outputs. It provides a flexible
+    way to use different LLM providers for scoring purposes.
+
+    We are using litellm to support multiple LLM providers.
+
+    Attributes:
+        model_id: The specific model identifier to use for scoring
+        temperature: Controls randomness in the LLM's responses (0.0 to 1.0)
+        max_tokens: Maximum number of tokens allowed in the LLM's response
+    """
+
+    model_id: str = Field(
+        description="The model to use, check https://docs.litellm.ai/docs/providers for supported models"
+    )
+    temperature: float = Field(
+        ..., description="The temperature to use for the response"
+    )
+    max_tokens: int = Field(
+        ..., description="The maximum number of tokens in the response"
+    )
+
+    # TODO: check if we can validate the model_id with litellm on a post_init method
+
+
+class InstructorLLMScorer(LLMScorer):
+    def __new__(cls, *args, **kwargs):  # type: ignore
+        raise DeprecationWarning(
+            "InstructorLLMScorer is deprecated and will be removed in a future version. "
+            "Use LLMScorer directly instead, which now has built-in support for structured outputs."
+        )
+
+
+class HuggingFacePipelineScorer(weave.Scorer):
+    """
+    Base class for using Hugging Face pipelines for moderation scoring.
+
+    This class simplifies the use of Hugging Face pipelines by handling the initialization and providing a common interface for scoring.
+
+    Args:
+        task (str): The pipeline task type (e.g., `"text-classification"`).
+        model_name_or_path (str): The name or path of the model to use.
+        device (str): The device to use for inference. Defaults to `"auto"`.
+
+    Example:
+        >>> from weave.scorers.moderation_scorer import PipelineScorer
+        >>> scorer = PipelineScorer(
+        ...     task="text-classification",
+        ...     model_name_or_path="distilbert-base-uncased-finetuned-sst-2-english",
+        ...     device="auto"
+        ... )
+        >>> output = scorer.pipe("This is a great movie!")
+        >>> print(output)
+        [{'label': 'POSITIVE', 'score': 0.9998}]
+    """
+
+    task: str = Field(
+        description="The task to use for the pipeline, for example 'text-classification'"
+    )
+    model_name_or_path: str = Field(default="", description="The path to the model")
+    device: Union[str, Any] = Field(
```

Comment:
This breaks serialization -- instead, you can do `device: str`, then make a property `_device: torch.device` that does the set_device stuff